lightning-invoice/src/lib.rs

```diff
@@ -1295,10 +1290,100 @@ impl<S> Display for SignOrCreationError<S> {
 	}
 }
 
+mod utils {
+	use {CreationError, Currency, Invoice, InvoiceBuilder};
+	use bitcoin_hashes::Hash;
+	use lightning::chain;
+	use lightning::chain::chaininterface::{BroadcasterInterface, FeeEstimator};
+	use lightning::chain::keysinterface::{Sign, KeysInterface};
+	use lightning::ln::channelmanager::{ChannelManager, MIN_FINAL_CLTV_EXPIRY};
+	use lightning::ln::features::InvoiceFeatures;
+	use lightning::routing::network_graph::RoutingFees;
+	use lightning::routing::router::RouteHintHop;
+	use lightning::util::logger::Logger;
+	use secp256k1::Secp256k1;
+	use std::ops::Deref;
+
+	/// Utility to construct an invoice.
+	#[allow(dead_code)]
+	pub fn from_channelmanager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>(
+		channelmanager: &ChannelManager<Signer, M, T, K, F, L>, amt_msat: Option<u64>, description: String,
+		network: Currency,
+	) -> Result<Invoice, CreationError>
+	where
+		M::Target: chain::Watch<Signer>,
+		T::Target: BroadcasterInterface,
+		K::Target: KeysInterface<Signer = Signer>,
+		F::Target: FeeEstimator,
+		L::Target: Logger,
+	{
+		// Marshall route hints.
+		let our_channels = channelmanager.list_usable_channels();
```

Comment:
IIUC this utility returns as routing hints the complete set of our channels, including the non-announced subset one ? 

Sounds pretty harmful for privacy, maybe the utility should have a boolean `all` or otherwise a Vec<short_channel_id> to precise which ones to disclose ?

Reply:
I think this is the right default to this - in most cases if we're a "client" node we only have one or two channels anyway (and in non-"client" cases we have several channels that are public), and the "real" fix is to generate random short channel ids.

Reply:
Agree for the "client" node but for a big hub you will disclose all your private channels with spoke *if* you're also a merchant which is quite the case of a bunch of such nodes. Though opening an issue for now, we'll adjust in function of users feedback.

Reply:
Right, we ultimately need some heuristic of "we have > 5 public channels, we shouldn't include any route hints at all", probably.